### PR TITLE
[FLINK-15912][table] Add Context to TableSourceFactory and TableSinkFactory

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Discussion in: http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/DISCUSS-Improve-TableFactory-td36647.html

Vote in: http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/VOTE-Improve-TableFactory-to-add-Context-td37211.html

Motivation:
Now the main needs and problems are:

Connector can't get TableConfig[1], and some behaviors really need to be
controlled by the user's table configuration. In the era of catalog, we
can't put these config in connector properties, which is too inconvenient.
A context class also allows for future modifications without touching the TableFactory interface again.

## Brief change log

1. Interface modification for table factories
2. Apply new interface in blink planner.

## Verifying this change

`SinkTest.testTableSourceSinkFactory`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableSinkFactory.java [hunk 1]
```diff
@@ -53,4 +55,39 @@
...
+		 */
+		ObjectIdentifier getObjectIdentifier();
+
+		/**
+		 * @return table {@link CatalogTable} instance.
+		 */
+		CatalogTable getTable();
+
+		/**
+		 * @return readable config of this table environment.
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableSinkFactory.java]: Add a description that the configuration gives the factory instance the ability to access `TableConfig#getConfiguration()` which holds the current TableEnvironment session configurations.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableSinkFactory.java [hunk 2]
```diff
@@ -53,4 +55,39 @@
...
 	}
 
+	/**
+	 * Creates and configures a {@link TableSink} based on the given
+	 {@link Context}.
+	 *
+	 * @param context context of this table sink.
+	 * @return the configured table sink.
+	 */
+	default TableSink<T> createTableSink(Context context) {
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableSinkFactory.java]: Should we deprecate the other `createTableSink` and `createTableSource` interfaces?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java [hunk 3]
```diff
@@ -103,4 +103,15 @@
 		return Optional.empty();
 	}
 
+	/**
```
Comment (wuchong) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java]: We should make sure `TableSinkFactory#createTableSink(context)` is the only entry to be invoked by the planner. However,  the above `createTableSinkForCatalogTable` method still calling the old `createTableSink` method in old planner.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SinkTest.scala [hunk 4]
```diff
@@ -70,24 +69,82 @@ class SinkTest extends TableTestBase {
...
-      ArgumentMatchers.any(), ArgumentMatchers.any())).thenReturn(sink)
-    util.tableEnv.registerCatalog(catalog.getName, catalog)
-    util.tableEnv.useCatalog(catalog.getName)
-    val catalogTable = new CatalogTableImpl(schema, Map[String, String]().asJava, "")
-    catalog.createTable(new ObjectPath("default", "tbl"), catalogTable, false)
-    util.tableEnv.sqlQuery("select 1").insertInto("tbl")
+  def testTableSourceSinkFactory(): Unit = {
+    val factory = new TestContextTableFactory
+    util.tableEnv.getConfig.getConfiguration.setBoolean(factory.needContain, true)
+    util.tableEnv.registerCatalog("cat", new GenericInMemoryCatalog("default") {
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SinkTest.scala]: This only guarantees one of the paths, i.e. via `TableFactoryUtil#createTableSinkForCatalogTable`, which is only used by Hive. However, the another path is not covered, i.e. via `CatalogSourceTable#findAndCreateTableSource` which is used by most users.

Maybe you can upgrade the existing `TestCollectionTableFactory` to use the new interface and expose a session configuration, e.g. `collection.is-bounded`, the flag will be accessed via `Context#getConfiguration` and pass to the created `CollectionTableSource`.

Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SinkTest.scala]: Yes, you are right, but I want to modify this step by step.
The previous method `createTableSink(ObjectPath tablePath, CatalogTable table)` is only work in `createTableSinkForCatalogTable` too.
after these commits looks good to you, I will create following commit to fix this.
What do you think?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java (hunk 3 above)
Comment (JingsongLi) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java]: I'll push another commit to `Support create table source/sink by context in legacy planner`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableSinkFactory.java (hunk 2 above)
Comment (JingsongLi) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/TableSinkFactory.java]: I don't know... we can deprecate `createTableSink(ObjectPath, CatalogTable)`. CC: @twalthr

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SinkTest.scala (hunk 4 above)
Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SinkTest.scala]: But without a test to cover the new interface, we have no idea whether it works as expected.

Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SinkTest.scala]: What do you mean? Is this not a test cover?

Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SinkTest.scala]: As I mentioned above, this test doesn't cover another path, i.e. `CatalogSourceTable#findAndCreateTableSource`.

Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/SinkTest.scala]: I'll modify these things by multi commits.

**Comment by JingsongLi** [comment]
We can wait this one: https://github.com/apache/flink/pull/11055

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/StreamTableSourceFactory.java [hunk 5]
```diff
@@ -39,7 +39,20 @@
...
+	}
+
+	/**
+	 * Creates and configures a {@link StreamTableSource} based on the given
+	 {@link Context}.
+	 *
+	 * @param context context of this table source.
+	 * @return the configured table source.
+	 */
+	default StreamTableSource<T> createStreamTableSource(Context context) {
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/StreamTableSourceFactory.java]: Somebody implements `StreamTableSourceFactory` looks like need this method.

[the remark this answers is not in the document]
Comment (wuchong) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/StreamTableSourceFactory.java]: But he can still use the `TableSource<T> createTableSource(context)` interface.

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/StreamTableSourceFactory.java [hunk 6]
```diff
@@ -48,4 +61,12 @@
 	default TableSource<T> createTableSource(Map<String, String> properties) {
 		return createStreamTableSource(properties);
 	}
+
+	/**
+	 * Only create a stream table source.
+	 */
+	@Override
+	default TableSource<T> createTableSource(Context context) {
+		return createStreamTableSource(context);
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/StreamTableSourceFactory.java]: indicate users to implement `createStreamTableSource(context)`?

File: flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/StreamTableSourceFactory.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/StreamTableSourceFactory.java]: if user use `TableSource`, there is no need to use `StreamTableSourceFactory`..

[the remark this answers is not in the document]
Comment (wuchong) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/StreamTableSourceFactory.java]: Yes. `StreamTableSourceFactory` should be deprecated. It is confused why there is a `StreamTableSourceFactory` and `TableSourceFactory`.  But for compatibility, users can still use `StreamTableSourceFactory` but indicate users to use `TableSource<T> createTableSource(context)`.

[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/factories/StreamTableSourceFactory.java]: OK, I will remove this method.

**Pull request closed without merging**